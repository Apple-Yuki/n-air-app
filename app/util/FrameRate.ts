// Utility functions for working with frame rate values

import _ from 'lodash';

type Rational = {
  denominator: number,
  numerator: number,
};

export default {

  rationalToFrameInterval(rational: Rational) {
    return rational.denominator / rational.numerator;
  },

  rationalToFramesPerSecond(rational: Rational) {
    return rational.numerator / rational.denominator;
  },

  // Operates on rational values. Determines if the rate
  // is within the given range.
  rateInRange(rate: Rational, min: Rational, max: Rational) {
    const t = this.rationalToFrameInterval(rate);
    const tMax = this.rationalToFrameInterval(min);
    const tMin = this.rationalToFrameInterval(max);

    return (tMin <= t) && (tMax >= t);
  },

  // Returns common FPS values which are valid in at least
  // one of the given FPS ranges.
  simpleFPSValuesForRanges(ranges: {min: Rational, max: Rational}[]) {
    return _.filter(this.commonFPSValues(), value => {
      return !!(_.find(ranges, range => {
        return this.rateInRange(value, range.min, range.max);
      }));
    });
  },

  // These are the same presets that the OBS UI uses
  commonFPSValues() {
    return [
      {
        name: '60',
        numerator: 60,
        denominator: 1
      },
      {
        name: '59.94',
        numerator: 60000,
        denominator: 1001
      },
      {
        name: '50',
        numerator: 50,
        denominator: 1
      },
      {
        name: '48',
        numerator: 48,
        denominator: 1
      },
      {
        name: '30',
        numerator: 30,
        denominator: 1
      },
      {
        name: '29.97',
        numerator: 30000,
        denominator: 1001
      },
      {
        name: '25',
        numerator: 25,
        denominator: 1
      },
      {
        name: '24',
        numerator: 24,
        denominator: 1
      },
      {
        name: '23.976',
        numerator: 24000,
        denominator: 1001
      }
    ];
  }

};
