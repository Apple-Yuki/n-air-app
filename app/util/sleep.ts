export function sleep (timeout: number) {
  return new Promise(done => setTimeout(done, timeout));
}
