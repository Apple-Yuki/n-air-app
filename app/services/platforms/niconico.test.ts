import { sleep } from 'util/sleep';
import { createSetupFunction } from 'util/test-setup';

const setup = createSetupFunction({
  injectee: {
    HostsService: {},
    SettingsService: {},
    UserService: {},
    StreamingService: {
      streamingStatusChange: {
        subscribe() {},
      },
    },
    WindowsService: {},
  },
});

beforeEach(() => {
  jest.resetModules();

  jest.mock('services/stateful-service');
  jest.mock('util/injector');
  jest.mock('services/streaming', () => ({}));
  jest.mock('services/user', () => ({}));
  jest.mock('services/settings', () => ({}));
  jest.mock('services/windows', () => ({}));
  jest.mock('util/sleep', () => ({
    sleep: () => sleep(0),
  }));
});

test('get instance', () => {
  setup();
  const { NiconicoService } = require('./niconico');
  expect(NiconicoService.instance).toBeInstanceOf(NiconicoService);
});

test('setupStreamSettingsで番組がない場合', async () => {
  setup();
  const { NiconicoService } = require('./niconico');
  const { instance } = NiconicoService;

  instance.fetchLiveProgramInfo = jest.fn().mockResolvedValue({});

  const result = await instance.setupStreamSettings();
  expect(result.url).toEqual('');
  expect(result.asking).toBe(false);
  expect(instance.fetchLiveProgramInfo).toHaveBeenCalledTimes(2);
});

test('setupStreamSettingsで番組がひとつある場合', async () => {
  const updatePlatformChannelId = jest.fn();
  const getSettingsFormData = jest.fn();
  const setSettings = jest.fn();

  getSettingsFormData.mockReturnValue([
    {
      nameSubCategory: 'Untitled',
      parameters: [{ name: 'service', value: '' }, { name: 'server', value: '' }, { name: 'key', value: '' }],
    },
  ]);

  const injectee = {
    UserService: {
      updatePlatformChannelId,
    },
    SettingsService: {
      getSettingsFormData,
      setSettings,
    },
  };

  setup({ injectee });
  const { NiconicoService } = require('./niconico');
  const { instance } = NiconicoService;

  instance.fetchLiveProgramInfo = jest.fn().mockResolvedValue({
    channelId: {
      url: 'url1',
      key: 'key1',
      bitrate: 'bitrate1',
    },
  });

  const result = await instance.setupStreamSettings();
  expect(result.url).toBe('url1');
  expect(updatePlatformChannelId).toHaveBeenCalledTimes(1);
  expect(updatePlatformChannelId).toHaveBeenLastCalledWith('channelId');
  expect(setSettings).toHaveBeenCalledTimes(1);
  expect(setSettings.mock.calls[0]).toMatchSnapshot();
});

test('setupStreamSettingsで番組取得にリトライで成功する場合', async () => {
  const updatePlatformChannelId = jest.fn();
  const getSettingsFormData = jest.fn();
  const setSettings = jest.fn();

  getSettingsFormData.mockReturnValue([
    {
      nameSubCategory: 'Untitled',
      parameters: [{ name: 'service', value: '' }, { name: 'server', value: '' }, { name: 'key', value: '' }],
    },
  ]);

  const injectee = {
    UserService: { updatePlatformChannelId: updatePlatformChannelId },
    SettingsService: { getSettingsFormData: getSettingsFormData, setSettings: setSettings },
  };

  setup({ injectee });
  const { NiconicoService } = require('./niconico');
  const { instance } = NiconicoService;

  instance.fetchLiveProgramInfo = jest
    .fn()
    .mockReturnValueOnce(Promise.resolve({}))
    .mockResolvedValue({
      channelId: {
        url: 'url1',
        key: 'key1',
        bitrate: 'bitrate1',
      },
    });

  const result = await instance.setupStreamSettings();
  expect(result).toEqual({
    asking: false,
    url: 'url1',
    key: 'key1',
    bitrate: 'bitrate1',
  });
  expect(updatePlatformChannelId).toHaveBeenCalledTimes(1);
  expect(updatePlatformChannelId).toHaveBeenLastCalledWith('channelId');
  expect(setSettings).toHaveBeenCalledTimes(1);
  expect(setSettings.mock.calls[0]).toMatchSnapshot();
});

test('setupStreamSettingsで番組が複数ある場合', async () => {
  const updatePlatformChannelId = jest.fn();
  const getSettingsFormData = jest.fn();
  const setSettings = jest.fn();
  const showWindow = jest.fn();

  getSettingsFormData.mockReturnValue([
    {
      nameSubCategory: 'Untitled',
      parameters: [{ name: 'service', value: '' }, { name: 'server', value: '' }, { name: 'key', value: '' }],
    },
  ]);

  const injectee = {
    UserService: {
      updatePlatformChannelId: updatePlatformChannelId,
    },
    SettingsService: {
      getSettingsFormData: getSettingsFormData,
      setSettings: setSettings,
    },
    WindowsService: {
      showWindow: showWindow,
    },
  };

  setup({ injectee });
  const { NiconicoService } = require('./niconico');
  const { instance } = NiconicoService;

  const info = {
    channelId1: { url: 'url1', key: 'key1', bitrate: 'bitrate1' },
    channelId2: { url: 'url2', key: 'key2', bitrate: 'bitrate2' },
  };
  instance.fetchLiveProgramInfo = jest.fn().mockResolvedValue(info);

  const result = await instance.setupStreamSettings();
  expect(result).toMatchInlineSnapshot(`
Object {
  "asking": true,
  "bitrate": undefined,
  "key": "",
  "url": "",
}
`);
  expect(showWindow.mock.calls[0]).toMatchSnapshot();

  expect(updatePlatformChannelId).not.toHaveBeenCalled();
  expect(setSettings).not.toHaveBeenCalled();
});
