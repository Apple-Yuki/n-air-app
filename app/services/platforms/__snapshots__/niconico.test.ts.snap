// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupStreamSettingsでストリーム情報がとれた場合 1`] = `
Array [
  "Stream",
  Array [
    Object {
      "nameSubCategory": "Untitled",
      "parameters": Array [
        Object {
          "name": "service",
          "value": "niconico ニコニコ生放送",
        },
        Object {
          "name": "server",
          "value": "url1",
        },
        Object {
          "name": "key",
          "value": "key1",
        },
      ],
    },
  ],
]
`;

exports[`setupStreamSettingsで番組取得にリトライで成功する場合 1`] = `
Array [
  "Stream",
  Array [
    Object {
      "nameSubCategory": "Untitled",
      "parameters": Array [
        Object {
          "name": "service",
          "value": "niconico ニコニコ生放送",
        },
        Object {
          "name": "server",
          "value": "url1",
        },
        Object {
          "name": "key",
          "value": "key1",
        },
      ],
    },
  ],
]
`;
